### PR TITLE
Withings

### PR DESCRIPTION
**Pull request by euguuu**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/gladys-4-docs) and the [website](https://documentation.gladysassistant.com).
- [x] Did you add fake requests data for the demo mode (`front/src/config/demo.json`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Service to integrate withings devices (by rest API + oauth2 generic implementation)

### CODE REVIEW

**Comment by euguuu** [comment]
improve init and poll, pr to test coverage

**Comment by euguuu** [comment]
fix prettier

**Comment by Pierre-Gilles** [comment]
Hi ! Congrats for this great PR.

I saw that you removed the WIP title of your PR, is it ready to review ?

If yes, make sure: 

- To run prettier on the frontend (currently failing)
- To fix merge conflict 

I'll review the PR once the CI is green 🙂

**Comment by euguuu** [comment]
this PR is ready

File: front/package.json [hunk 1]
```diff
@@ -53,6 +53,7 @@
     "qrcode": "^1.4.2",
     "react-big-calendar": "^0.22.1",
     "react-datepicker": "^2.13.0",
+    "react-router": "^5.2.0",
```
Comment (atrovato) [inline front/package.json]: Do you really need this? I don't see any usage. You use `preact-router`.

File: front/src/actions/dashboard/boxes/health.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import { RequestStatus } from '../../../utils/consts';
+import createBoxActions from '../boxActions';
+// import createDeviceActions from '../../device';
+import update from 'immutability-helper';
+import get from 'get-value';
+
+const BOX_KEY = 'Health';
+
+function createActions(store) {
```
Comment (atrovato) [inline front/src/actions/dashboard/boxes/health.js:9]: Why do you create this and not use `devicesInRoom.js` actions?
Health devices are not "classical devices"?

File: front/src/components/boxs/device-in-room/device-features/SensorDeviceFeature.jsx [hunk 3]
```diff
@@ -41,6 +41,11 @@ const SensorDeviceType = ({ children, ...props }) => (
             {props.deviceFeature.unit === DEVICE_FEATURE_UNITS.AMPERE && 'A'}
             {props.deviceFeature.unit === DEVICE_FEATURE_UNITS.VOLT && 'V'}
             {props.deviceFeature.unit === DEVICE_FEATURE_UNITS.PPM && 'ppm'}
+            {props.deviceFeature.unit === DEVICE_FEATURE_UNITS.KILOGRAM && 'kg'}
+            {props.deviceFeature.unit === DEVICE_FEATURE_UNITS.CENTIMETER && 'kg'}
```
Comment (atrovato) [inline front/src/components/boxs/device-in-room/device-features/SensorDeviceFeature.jsx]: Centimeter with kg?

File: front/src/components/boxs/device-in-room/device-features/SensorDeviceFeature.jsx [hunk 4]
```diff
@@ -41,6 +41,11 @@ const SensorDeviceType = ({ children, ...props }) => (
...
+            {props.deviceFeature.unit === DEVICE_FEATURE_UNITS.MERCURE_MILIMETER && 'kg'}
```
Comment (atrovato) [inline front/src/components/boxs/device-in-room/device-features/SensorDeviceFeature.jsx]: What is mercure milimeter?
Is it kg?

File: front/src/components/boxs/health/EditHealthBox.jsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  };
+
+  componentDidMount() {
+    this.getHealthData();
+  }
+
+  render(props, { selectedDeviceFeaturesOptions, deviceOptions, loading }) {
+    return (
+      <BaseEditBox {...props} titleKey="dashboard.boxTitle.health">
+        <div class={loading ? 'dimmer active' : 'dimmer'}>
```
Comment (atrovato) [inline front/src/components/boxs/health/EditHealthBox.jsx]: Use can use `classnames` to do this:
```
import cx from 'classnames';

<div class={cx('dimmer', { active: loading })}>
```

File: front/src/components/boxs/health/HealthBox.jsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  return (
+    <div class="card">
+      <div class="card-header">
+        <h3 class="card-title">{props.userName}</h3>
+      </div>
+      {props.loading && (
+        <div class="card-body o-auto" style={cardStyle}>
+          <div class={props.loading ? 'dimmer active' : 'dimmer'}>
+            <div class="loader" />
+            <div class="dimmer-content">{props.loading && <div style={minHeight} />}</div>
```
Comment (atrovato) [inline front/src/components/boxs/health/HealthBox.jsx:28]: I'm not really sur we like `style` usage.

File: front/src/components/oauth2/actions.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    store.setState({
+      secretId: e.target.value
+    });
+  },
+  async getCurrentConfig(state) {
+    store.setState({
+      oauth2GetStatus: RequestStatus.Getting
+    });
+
+    const returnServiceId = await state.httpClient.get(`/api/v1/service/${state.integrationName}/getServiceId`);
```
Comment (atrovato) [inline front/src/components/oauth2/actions.js:24]: `getServiceId` as REST path?

File: front/src/components/oauth2/actions.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    const returnGetConfig = await state.httpClient.get('/api/v1/service/oauth2/getCurrentConfig', {
```
Comment (atrovato) [inline front/src/components/oauth2/actions.js:26]: `getCurrentConfig` as REST path?

File: front/src/components/oauth2/actions.js [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    try {
+      if (!state.secretId || !state.clientId) {
+        store.setState({
+          oauth2GetStatus: RequestStatus.Error,
+          oauth2ErrorMsg: 'errorEmptyConfig'
+        });
+        return;
+      }
+
+      const returnSaveVar = await state.httpClient.post(`/api/v1/service/${state.integrationName}/savevar`, {
```
Comment (atrovato) [inline front/src/components/oauth2/actions.js:58]: What `savevar` means?

File: front/src/components/oauth2/actions.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      if (returnSaveVar.success === true) {
+        serviceId = returnSaveVar.result.serviceId;
+      } else {
+        store.setState({
+          oauth2GetStatus: RequestStatus.Error,
+          oauth2ErrorMsg: 'errorServiceId'
+        });
+      }
+
+      const returnValue = await state.httpClient.post('/api/v1/service/oauth2/buildAuthorizationUri', {
```
Comment (atrovato) [inline front/src/components/oauth2/actions.js:74]: Not really fan of REST path name `buildAuthorizationUri`

File: front/src/components/oauth2/actions.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      });
+      return;
+    }
+  },
+  async unConnect(state) {
+    store.setState({
+      oauth2GetStatus: RequestStatus.Getting
+    });
+
+    await state.httpClient.get(`/api/v1/service/${state.integrationName}/deleteConfig`);
```
Comment (atrovato) [inline front/src/components/oauth2/actions.js:102]: Calling GET to `deleteConfig`?

File: front/src/config/i18n/en.json [hunk 12]
```diff
@@ -661,6 +668,48 @@
...
+    },
+    "withings": {
+      "settings": {
+        "title": "Withings Settings",
+        "menu": "Settings",
+        "introduction": "To connect Withings account to Gladys, you first need to create a Withings user account in <a href=https://account.withings.com/connectionwou/account_login target=_blank >this link</a>.<br/><br/>To connect Withings measures to Gladys, you need to create a Withings partner account in <a href=https://account.withings.com/partner/add_oauth2 target=_blank >this link</a></br>(in callback url field put http://gladys.local/dashboard/integration/health/withings/settings, in environmnet field choose Dev, you can use Prod with gladys plus url and https protocol, reefer to Documentation link for more details). <br/>",
+        "instructionsToUse": "To use this integration, go to the dashboard and create a health box.",
+        "configurationError": "We were not able to configure your account. Are you sure your client_id and secret_id are valid?",
+        "complete": "Your account is now linked to Gladys with those parameters:",
+        "clientId": "Client Id: ",
```
Comment (atrovato) [inline front/src/config/i18n/en.json]: White space as last char.

File: front/src/config/i18n/en.json [hunk 13]
```diff
@@ -1208,6 +1257,25 @@
       "acceleration-z": "Acceleration Z",
       "bed-activity": "Bed Activity"
     },
+    "health": {
+      "battery": "Battery",
```
Comment (atrovato) [inline front/src/config/i18n/en.json]: Why do use the common `Battery` word?

File: front/src/config/i18n/en.json [hunk 14]
```diff
@@ -221,6 +222,12 @@
         "noImageToShow": "No image to show",
         "editBoxNameLabel": "Enter the name you want to give to the box:",
         "editBoxNamePlaceholder": "Name of the box"
+      },
+      "health": {
+        "editDeviceFeaturesLabel": "Select the devices you want to display.",
+        "noValue": "No value recorded",
+        "deviceTitle": "{{name}} - {{type}}",
+        "noDevices": "Looks you don't have any devices in your house!<br/>Define your device in the integration page."
```
Comment (atrovato) [inline front/src/config/i18n/en.json:230]: noDevices isn't in MarkupText, is the <br/> well interpreted?

File: front/src/config/i18n/fr.json [hunk 15]
```diff
@@ -221,6 +222,12 @@
         "noImageToShow": "Aucune image à afficher",
         "editBoxNameLabel": "Entrez le nom que vous souhaitez donner à la box :",
         "editBoxNamePlaceholder": "Nom de la box"
+      },
+      "health": {
+        "editDeviceFeaturesLabel": "Sélectionnez les appareils que vous voulez afficher ici.",
+        "noValue": "Aucune valeur enregistrée.",
+        "deviceTitle": "{{name}} - {{type}}",
+        "noDevices": "Vous n'avez pas d'appareil dans votre maison ! <br/> Définissez votre appareil dans la page de l'intégration."
```
Comment (atrovato) [inline front/src/config/i18n/fr.json:230]: Whitepaces not need around <br/>

File: front/src/config/i18n/fr.json [hunk 16]
```diff
@@ -661,6 +668,48 @@
         "leak": "Capteur de fuite",
         "magnet": "Capteur magnétique"
       }
+    },
+    "oauth2": {
+      "buttonConnect": "Connecter maintenant",
+      "apiKeyLabel": "Saisissez ensuite votre clé client_id et secret_id ici : ",
```
Comment (atrovato) [inline front/src/config/i18n/fr.json]: Can we remove the last whitepace char?

File: front/src/config/i18n/fr.json [hunk 17]
```diff
@@ -661,6 +668,48 @@
...
+      "settings": {
+        "title": "Withings Configuration",
+        "menu": "Configuration",
+        "introduction": "Pour connecter le compte Withings à Gladys, vous devez d'abord créer un compte d'utilisateur Withings à <a href=https://account.withings.com/connectionwou/account_login target=_blank >ce lien</a>.<br/><br/>Pour relier les mesures Withings à Gladys, vous devez créer un compte partenaire Withings à <a href=https://account.withings.com/partner/add_oauth2 target=_blank >ce lien</a></br>(dans le champ url de callback mettre http://gladys.local/dashboard/integration/health/withings/settings, dans le champs environnement choisir Dev, vous pouvez utiliser Prod avec gladys plus url et le protocole https, voir le lien Documentation pour plus de détails). <br/>",
+        "instructionsToUse": "Pour utiliser cette intégration, rendez-vous sur le tableau de bord et créez une box de santé.",
+        "configurationError": "Nous n'avons pas pu configurer votre compte. Êtes-vous sûr que votre client_id et votre secret_id sont corrects ?",
+        "complete": "Votre compte est maintenant lié à Gladys avec ces paramètres:",
+        "clientId": "Client Id: ",
```
Comment (atrovato) [inline front/src/config/i18n/fr.json]: Whitespace as last char.

File: front/src/routes/integration/all/withings/device-page/actions.js [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+  async getWithingsDevice(state) {
+    store.setState({
+      withingsGetStatus: RequestStatus.Getting
+    });
+    try {
+      const options = {};
+      const withingsDevicesReceived = await state.httpClient.get('/api/v1/service/withings/device', options);
+
+      // Build map of device image by device name
+      const mapOfDeviceImgByName = new Map();
```
Comment (atrovato) [inline front/src/routes/integration/all/withings/device-page/actions.js:48]: I think we prefer object instead of map?

Comment (atrovato) [reply front/src/routes/integration/all/withings/device-page/actions.js:48]: For more, this can be extract as constant.

File: package.json [hunk 19]
```diff
@@ -33,6 +33,7 @@
   "devDependencies": {
     "all-contributors-cli": "^6.15.0",
     "npm-run-all": "^4.1.5",
+    "nyc": "^15.1.0",
```
Comment (atrovato) [inline package.json]: I can't see any usage, do we really need this?

File: server/api/controllers/health.controller.js [hunk 20]
```diff
@@ -0,0 +1,78 @@
+const asyncMiddleware = require('../middlewares/asyncMiddleware');
+// const { Error400 } = require('../../utils/httpErrors');
+// const { ERROR_MESSAGES } = require('../../utils/constants');
```
Comment (atrovato) [inline server/api/controllers/health.controller.js]: No needed comments.

**Review by atrovato — COMMENTED** [review]
Ok, I only review 20 files... that's a big PR.
But I see you add OAuth2 management (dupe https://github.com/GladysAssistant/Gladys/pull/522).
Maybe we can separate both OAuth2 and Withings features?

File: front/package.json (hunk 1 above)
Comment (euguuu) [reply front/package.json]: it's a mistake , I remove this

File: front/src/actions/dashboard/boxes/health.js (hunk 2 above)
Comment (euguuu) [reply front/src/actions/dashboard/boxes/health.js:9]: this box search only device feature with category=health and health device is not necessary linked to a room (for example a watch is not linked to a specific room, I don't see this capability in gladys now but in next version I want suggest to add capability to linked a device to an user like  a room (some withings devices can be linked to different users, like scale, for now to change withings users it's necessary to make 2 separate gladys user and I think must be cool to put this choicein box display like 'présence')

File: front/src/components/boxs/device-in-room/device-features/SensorDeviceFeature.jsx (hunk 3 above)
Comment (euguuu) [reply front/src/components/boxs/device-in-room/device-features/SensorDeviceFeature.jsx]: oups ... it's mistake, I change this

File: front/src/components/boxs/device-in-room/device-features/SensorDeviceFeature.jsx (hunk 4 above)
Comment (euguuu) [reply front/src/components/boxs/device-in-room/device-features/SensorDeviceFeature.jsx]: it's presure unitused by very specific withings device (blood presure)
the good unit is mmHg I change this (same mistake of before comment)

File: front/src/components/boxs/health/EditHealthBox.jsx (hunk 5 above)
Comment (euguuu) [reply front/src/components/boxs/health/EditHealthBox.jsx]: ok I change this, but I just copy what is done in EditDeviceInRoom.js

File: front/src/config/i18n/fr.json (hunk 16 above)
Comment (euguuu) [reply front/src/config/i18n/fr.json]: I remove it

File: front/src/config/i18n/fr.json (hunk 17 above)
Comment (euguuu) [reply front/src/config/i18n/fr.json]: I remove it

File: package.json (hunk 19 above)
Comment (euguuu) [reply package.json]: It's a mistake I remove this

File: server/api/controllers/health.controller.js (hunk 20 above)
Comment (euguuu) [reply server/api/controllers/health.controller.js]: I remove it

**Comment by euguuu** [comment]
for oatuh2 , 
in my comprehension #522 is feature to add oauth2 capability to gladys (other app want connect to gladys)
in this PR my oauth2 implementation is the other side : gladys want communicate with third party

I am wrong ?

File: front/src/routes/integration/all/withings/device-page/actions.js (hunk 18 above)
Comment (euguuu) [reply front/src/routes/integration/all/withings/device-page/actions.js:48]: I'm not sure to understand, 
you talk about a list of object with 2 attributes(device name / img name) ? If I do that I need make a function to search element in the list and a map sounds more performant for that, no ? (I'm not a good js dev so maybe better solution exist)

**Comment by atrovato** [comment]
You're right ;)

File: front/src/components/boxs/health/HealthBox.jsx (hunk 6 above)
Comment (euguuu) [reply front/src/components/boxs/health/HealthBox.jsx:28]: In fact I just copy DevicesInRoomBox

File: front/src/components/oauth2/actions.js (hunk 8 above)
Comment (euguuu) [reply front/src/components/oauth2/actions.js:26]: I use it to get client_id and secret store in table t_variable, I need it to send authorisation link to third party authentication page and when this party back to callbackUrl choose the good init method (normally oauth2 page can be used for all others oauth2  provider just by config, only initialisation and poll method must be implemented by a new gladys service)

File: front/src/components/oauth2/actions.js (hunk 7 above)
Comment (euguuu) [reply front/src/components/oauth2/actions.js:24]: I use it to get client_id and secret store in table t_variable, I need it to send authorisation link to third party authentication page and when this party back to callbackUrl choose the good init method (normally oauth2 page can be used for all others oauth2 provider just by config, only initialisation and poll method must be implemented by a new gladys service)

File: front/src/components/oauth2/actions.js (hunk 9 above)
Comment (euguuu) [reply front/src/components/oauth2/actions.js:58]: it's to save variable in t_variable table with serviceId column

File: front/src/components/oauth2/actions.js (hunk 10 above)
Comment (euguuu) [reply front/src/components/oauth2/actions.js:74]: ok have you a suggestion for new name? I can change it fastly

File: front/src/components/oauth2/actions.js (hunk 11 above)
Comment (euguuu) [reply front/src/components/oauth2/actions.js:102]: no param pass to querry, only integration name in path is used if you prefer I put it in post

File: front/src/config/i18n/en.json (hunk 12 above)
Comment (euguuu) [reply front/src/config/i18n/en.json]: I remove it

File: front/src/config/i18n/en.json (hunk 13 above)
Comment (euguuu) [reply front/src/config/i18n/en.json]: I change it by deviceFeatureCategory.battery.shortCategoryName

File: front/src/config/i18n/fr.json (hunk 15 above)
Comment (euguuu) [reply front/src/config/i18n/fr.json:230]: I remove it

File: front/src/config/i18n/en.json (hunk 14 above)
Comment (euguuu) [reply front/src/config/i18n/en.json:230]: yes I use it in <Text> and works fine

File: server/package.json [hunk 21]
```diff
@@ -87,14 +87,19 @@
     "joi": "^14.3.1",
     "jsonwebtoken": "^8.4.0",
     "lodash.clonedeep": "^4.5.0",
+    "mock-http-server": "^1.4.2",
```
Comment (Pierre-Gilles) [inline server/package.json]: Where is it used?

File: server/package.json [hunk 22]
```diff
@@ -73,8 +72,9 @@
     "@gladysassistant/gladys-gateway-js": "^3.2.6",
     "@hapi/joi": "^17.1.0",
     "@hapi/joi-date": "^2.0.1",
+    "axios": "^0.19.2",
```
Comment (Pierre-Gilles) [inline server/package.json]: If you use specific dependency in your service, please add them to your service package.json

File: server/package.json [hunk 23]
```diff
@@ -87,14 +87,19 @@
     "joi": "^14.3.1",
     "jsonwebtoken": "^8.4.0",
     "lodash.clonedeep": "^4.5.0",
+    "mock-http-server": "^1.4.2",
     "node-nlp": "^3.0.3",
     "node-schedule": "^1.3.2",
     "node-webcrypto-ossl": "^1.0.48",
+    "nyc": "^15.1.0",
+    "oauth2-mock-server": "^1.5.1",
```
Comment (Pierre-Gilles) [inline server/package.json]: If you add dev-dependencies, please add them to the devDependencies not the dependencies

Why do you need nyc btw?

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Thanks for your PR ! :) 

First, you added your development.db file to this PR, please remove it! I don't even know how you did this as it's in the .gitignore... ^^

Then, I added a few feedbacks on dependencies. 

Once you removed the .db file I will be able to test it more extensively.

File: server/package.json (hunk 22 above)
Comment (euguuu) [reply server/package.json]: this dependancy is not use in withings service but in server/lob/oauth2/executeQuery.js

File: server/package.json (hunk 23 above)
Comment (euguuu) [reply server/package.json]: I remove nyc (it's a mistake whenI try to execute coverage in my computer, I don't understand ho I can run it, I always push PR to have the coverage ...)

oauth2-mock-server is used in oauth2 test, I put it in devDependencies

File: server/package.json (hunk 21 above)
Comment (euguuu) [reply server/package.json]: mock-http-server is used in withings test, I put it in devDependencies

File: server/package.json (hunk 23 above)
Comment (euguuu) [reply server/package.json]: I remove db file

Comment (euguuu) [reply server/package.json]: If you have no withings account I can give you my client & secret id for test

Comment (cicoub13) [reply server/package.json]: > I remove nyc (it's a mistake whenI try to execute coverage in my computer, I don't understand ho I can run it, I always push PR to have the coverage ...)

You can execute it with `npm run coverage` or `./node_modules/.bin/nyc --reporter=html npm run test-service --service=withings`
The second command will generate a `coverage` folder with report inside (`index.html`)

Comment (euguuu) [reply server/package.json]: many thanks

**Comment by euguuu** [comment]
@Pierre-Gilles can you say me if you see this pr in your review list? (i merge master and resolve your requested change)

**Comment by Pierre-Gilles** [comment]
No I still don't see it! 

This is what I see: (Changes requested)
Screenshot 2021-01-15 at 10 05 38

Weird

**Comment by euguuu** [comment]
@Pierre-Gilles 
I don't know how I can change this, I resolve ll change requested and re request a review

image

**Review by euguuu — COMMENTED** [review]
I remove dbfile and fix bad dependancies

File: server/utils/constants.js [hunk 24]
```diff
@@ -439,6 +467,7 @@ const DASHBOARD_BOX_TYPE = {
   USER_PRESENCE: 'user-presence',
   CAMERA: 'camera',
   DEVICES_IN_ROOM: 'devices-in-room',
+  HELTH: 'health',
```
Comment (VonOx) [inline server/utils/constants.js]: @euguuu little typo

Comment (euguuu) [reply server/utils/constants.js]: I fix this error

**Pull request closed without merging**